File: Impostor/SecretWordViewController.m
//
//  SecretWordViewController.m
//  Impostor
//
//  Created by William Entriken on 12/14/13.
//  Copyright (c) 2013 William Entriken. All rights reserved.
//

#import "Impostor-Swift.h"
#import "SecretWordViewController.h"
#import <AudioToolbox/AudioToolbox.h>
#import "SCLAlertView.h"

@interface SecretWordViewController () <UIImagePickerControllerDelegate,UINavigationControllerDelegate>
@property (nonatomic) ImpostorGameModel *game;
@property (nonatomic) SCLAlertView *sclAlertView;
@property (nonatomic) UIImagePickerController *imagePickerController;
@property (nonatomic) UIPopoverController *imagePopoverController;
@property (nonatomic) BOOL wantsToTakePhoto;
@property (nonatomic) BOOL photoDenied;
@end

@implementation SecretWordViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
    self.game = [ImpostorGameModel game];
    self.topSecretLabel.transform = CGAffineTransformMakeRotation(-10 * M_PI / 180.0);
    self.topSecretLabel.clipsToBounds = NO;
}

- (IBAction)showSecretWord:(id)sender {
    SystemSoundID soundID;
    NSURL *url = [[NSBundle mainBundle] URLForResource:@"peek" withExtension:@"mp3"];
    AudioServicesCreateSystemSoundID((__bridge CFURLRef)url, &soundID);
    AudioServicesPlaySystemSound (soundID);
    NSString *theWord = (self.game.playerWords)[self.playerNumber];
    
    self.sclAlertView = [[SCLAlertView alloc] init];
    self.sclAlertView.shouldDismissOnTapOutside = YES;
    self.sclAlertView.backgroundType = Blur;
    self.sclAlertView.labelTitle.font = [UIFont fontWithName:@"Chalkboard SE" size:20.0];
    self.sclAlertView.viewText.font = [UIFont fontWithName:@"Chalkboard SE" size:16.0];
    UIImage *appIcon = [UIImage imageNamed:@"AppIcon60x60"];
    [self.sclAlertView showCustom:self image:appIcon color:[UIColor blackColor] title:NSLocalizedString(@"Your secret word is", @"On the secret word screen") subTitle:theWord closeButtonTitle:NSLocalizedString(@"OK",@"Dismiss the popup") duration:0.0f];
    
    [self.sclAlertView alertIsDismissed:^{
        if (self.playerNumber == self.game.numberOfPlayers-1) {
            [self.game doneShowingSecretWords];
            NSArray *viewControllers = [self.navigationController viewControllers];
            [self.navigationController popToViewController:viewControllers[1] animated:YES];
            return;
        }
        
        UIStoryboard *storyboard;
        storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
        SecretWordViewController *newController = [storyboard instantiateViewControllerWithIdentifier:@"secretWord"];
        newController.playerNumber = self.playerNumber+1;
        [self.navigationController pushViewController:newController animated:YES];
    }];
}

- (IBAction)stopGame:(id)sender {
    [self.navigationController popToRootViewControllerAnimated:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    self.playerLabel.text = [NSString stringWithFormat:NSLocalizedString(@"Player #%ld", @"Current player"), (long)self.playerNumber+1];
    UIImage *photo = [UIImage imageNamed:@"defaultHeadshot.png"];
    self.playerImage.image = photo;
    
    NSString *imageName = [NSString stringWithFormat:@"%ld", (long)self.playerNumber];
    photo = [[CachedPersistentJPEGImageStore sharedStore] imageWithName:imageName];

    if (photo)
        self.playerImage.image = photo;
    self.wantsToTakePhoto = !photo;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    self.screenName = @"SecretWordViewController";
    
    if (!self.wantsToTakePhoto || self.photoDenied)
        return;
    if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
        self.imagePickerController = [[UIImagePickerController alloc] init];
        self.imagePickerController.delegate = self;
        self.imagePickerController.sourceType = UIImagePickerControllerSourceTypeCamera;
        self.imagePickerController.mediaTypes = [UIImagePickerController availableMediaTypesForSourceType:
                                                 UIImagePickerControllerSourceTypeCamera];
        if([UIImagePickerController isCameraDeviceAvailable:UIImagePickerControllerCameraDeviceFront])
            self.imagePickerController.cameraDevice = UIImagePickerControllerCameraDeviceFront;
        
        [self presentViewController:self.imagePickerController animated:YES completion:nil];
    }
/*
    else {
        // THIS IS JUST FOR IPODS AND THE SIMULATOR
        self.imagePickerController = [[UIImagePickerController alloc] init];
        self.imagePickerController.delegate = self;
        self.imagePickerController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
        // On iPad use pop-overs.
        if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
            self.imagePopoverController = [[UIPopoverController alloc] initWithContentViewController:self.imagePickerController];
            [self.imagePopoverController presentPopoverFromRect:self.topSecretLabel.frame
                                     inView:self.view
                   permittedArrowDirections:UIPopoverArrowDirectionAny
                                   animated:YES];
        }
        else {
            // On iPhone use full screen presentation.
            [[self presentingViewController] presentViewController:self.imagePickerController animated:YES completion:nil];
        }
    }
*/
}

#pragma mark - UIImagePickerControllerDelegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info
{
    UIImage *photo = info[UIImagePickerControllerOriginalImage];
    // http://stackoverflow.com/questions/5427656/ios-uiimagepickercontroller-result-image-orientation-after-upload/5427890#5427890
    UIImage *normalizedImage;
    if (photo.imageOrientation == UIImageOrientationUp)
        normalizedImage = photo;
    else {
        UIGraphicsBeginImageContextWithOptions(photo.size, NO, photo.scale);
        [photo drawInRect:(CGRect){0, 0, photo.size}];
        normalizedImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
    }
    normalizedImage = [[self class] cropBiggestCenteredSquareImageFromImage:normalizedImage withSide:800];
    
    NSString *imageName = [NSString stringWithFormat:@"%ld", (long)self.playerNumber];
    [[CachedPersistentJPEGImageStore sharedStore] saveImage:normalizedImage withName:imageName];
    self.playerImage.image = normalizedImage;
    
    [picker dismissViewControllerAnimated:YES completion:nil];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    self.photoDenied = YES;
    [picker dismissViewControllerAnimated:YES completion:nil];
}

// http://stackoverflow.com/questions/14917770/finding-the-biggest-centered-square-from-a-landscape-or-a-portrait-uiimage-and-s
+ (UIImage*) cropBiggestCenteredSquareImageFromImage:(UIImage*)image withSide:(CGFloat)side
{
    // Get size of current image
    CGSize size = [image size];
    if( size.width == size.height && size.width == side){
        return image;
    }
    
    CGSize newSize = CGSizeMake(side, side);
    double ratio;
    double delta;
    CGPoint offset;
    
    //make a new square size, that is the resized imaged width
    CGSize sz = CGSizeMake(newSize.width, newSize.width);
    
    //figure out if the picture is landscape or portrait, then
    //calculate scale factor and offset
    if (image.size.width > image.size.height) {
        ratio = newSize.height / image.size.height;
        delta = ratio*(image.size.width - image.size.height);
        offset = CGPointMake(delta/2, 0);
    } else {
        ratio = newSize.width / image.size.width;
        delta = ratio*(image.size.height - image.size.width);
        offset = CGPointMake(0, delta/2);
    }
    
    //make the final clipping rect based on the calculated values
    CGRect clipRect = CGRectMake(-offset.x, -offset.y,
                                 (ratio * image.size.width),
                                 (ratio * image.size.height));
    
    //start a new context, with scale factor 0.0 so retina displays get
    //high quality image
    if ([[UIScreen mainScreen] respondsToSelector:@selector(scale)]) {
        UIGraphicsBeginImageContextWithOptions(sz, YES, 0.0);
    } else {
        UIGraphicsBeginImageContext(sz);
    }
    UIRectClip(clipRect);
    [image drawInRect:clipRect];
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return newImage;
}


@end

File: Impostor/SecretWordViewController.swift
//
//  SecretWordViewController.swift
//  Impostor
//
//  Created by William Entriken on 4/24/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import UIKit
import AudioToolbox
import CDAlertView
import Firebase
import MobileCoreServices

class SecretWordViewController: UIViewController {
    @IBOutlet weak var playerLabel: UILabel!
    @IBOutlet weak var playerImage: UIImageView!
    @IBOutlet weak var topSecretLabel: UILabel!
    var playerNumber: Int!
    
    fileprivate var game: ImpostorGameModel!
    fileprivate var imagePickerController: UIImagePickerController!
    fileprivate var wantsToTakePhoto = true
    fileprivate var photoDenied = false
    fileprivate var accentSoundId: SystemSoundID = {
        let url = Bundle.main.url(forResource: "peek", withExtension: "mp3")!
        var soundID: SystemSoundID = 0
        AudioServicesCreateSystemSoundID(url as CFURL, &soundID)
        return soundID
    }()

    
    override func viewDidLoad() {
        super.viewDidLoad()
        game = ImpostorGameModel.game
        topSecretLabel.transform = CGAffineTransform(rotationAngle: -10 * CGFloat(Double.pi) / 180.0)
        topSecretLabel.clipsToBounds = false
    }
    
    @IBAction func showSecretWord(_ sender: AnyObject) {
        AudioServicesPlaySystemSound(accentSoundId);
        let theWord = game.playerWords[playerNumber]
        
        let alertView = CDAlertView(title: NSLocalizedString("Your secret word is", comment: "On the secret word screen"),
                                    message: theWord,
                                    type: .custom(image: UIImage(named:"AppIcon60x60")!))
        alertView.alertBackgroundColor = UIColor.black
        alertView.titleTextColor = UIColor.white
        alertView.messageTextColor = UIColor.white
        alertView.titleFont = UIFont(name: "AmericanTypewriter", size: 20.0)!
        alertView.messageFont = UIFont(name: "AmericanTypewriter-Bold", size: 20.0)!
        
        let dismissBlock: (CDAlertViewAction) -> Bool = {_ in
            if self.playerNumber == self.game.numberOfPlayers - 1 {
                self.game.doneShowingSecretWords()
                var viewControllers: [AnyObject] = self.navigationController!.viewControllers
                self.navigationController!.popToViewController(viewControllers[1] as! UIViewController, animated: true)
                return true
            }
            let storyboard = UIStoryboard(name: "Main", bundle: nil)
            let newController = storyboard.instantiateViewController(withIdentifier: "secretWord") as! SecretWordViewController
            newController.playerNumber = self.playerNumber + 1
            self.navigationController!.pushViewController(newController, animated: true)
            return true
        }
        let action = CDAlertViewAction(title: NSLocalizedString("OK", comment: "Dismiss the popup"),
                                       font: UIFont(name: "AmericanTypewriter-Bold", size: 20.0)!,
                                       textColor: UIColor.white,
                                       backgroundColor: nil, handler: dismissBlock)
        alertView.add(action: action)
        alertView.show()
    }
    
    @IBAction func stopGame(_ sender: AnyObject) {
        self.navigationController!.popToRootViewController(animated: true)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        Analytics.logEvent(AnalyticsEventViewItem, parameters: [
            AnalyticsParameterContentType:"view" as NSObject,
            AnalyticsParameterItemID:NSStringFromClass(type(of: self)) as NSObject
            ])
        
        self.playerLabel.text = String(format: NSLocalizedString("Player #%ld", comment: "Current player"), self.playerNumber + 1)
        
        let imageName = "\(Int(self.playerNumber))"
        if let photo = CachedPersistentJPEGImageStore.sharedStore.imageWithName(imageName) {
            self.playerImage.image = photo
            self.wantsToTakePhoto = false
        } else {
            self.playerImage.image = UIImage(named: "defaultHeadshot.png")!
            self.wantsToTakePhoto = true
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        guard self.wantsToTakePhoto && !self.photoDenied else {
            return
        }
        guard UIImagePickerController.isSourceTypeAvailable(.camera) else {
            return
        }
        imagePickerController = UIImagePickerController()
        imagePickerController.delegate = self
        imagePickerController.sourceType = .camera
        imagePickerController.mediaTypes = [kUTTypeImage as String]
        if UIImagePickerController.isCameraDeviceAvailable(.front) {
            imagePickerController.cameraDevice = .front
        }
        imagePickerController.modalPresentationStyle = .fullScreen
        present(imagePickerController, animated: true, completion: nil)
    }
    
}

extension SecretWordViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        var photo = info[.originalImage] as! UIImage
        photo = photo.normalizedImage().cropBiggestCenteredSquareImage(withSide: 800)
        let imageName: String = "\(Int(self.playerNumber))"
        CachedPersistentJPEGImageStore.sharedStore.saveImage(photo, withName: imageName)
        self.playerImage.image = photo
        picker.dismiss(animated: true, completion: nil)
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        self.photoDenied = true
        picker.dismiss(animated: true, completion: nil)
    }
}

private extension UIImage {
    func normalizedImage() -> UIImage {
        if (self.imageOrientation == UIImage.Orientation.up) {
            return self;
        }
        
        UIGraphicsBeginImageContextWithOptions(self.size, false, self.scale);
        let rect = CGRect(x: 0, y: 0, width: self.size.width, height: self.size.height)
        self.draw(in: rect)
        
        let normalizedImage : UIImage = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext();
        return normalizedImage;
    }
    
    func cropBiggestCenteredSquareImage(withSide side: CGFloat) -> UIImage {
        if self.size.height == side && self.size.width == side {
            return self
        }
        let newSize = CGSize(width: side, height: side)
        let ratio: CGFloat
        let delta: CGFloat
        let offset: CGPoint
        if self.size.width > self.size.height {
            ratio = newSize.height / self.size.height
            delta = ratio * (self.size.width - self.size.height)
            offset = CGPoint(x: delta / 2, y: 0)
        }
        else {
            ratio = newSize.width / self.size.width
            delta = ratio * (self.size.height - self.size.width)
            offset = CGPoint(x: 0, y: delta / 2)
        }
        let clipRect = CGRect(x: -offset.x, y: -offset.y, width: ratio * self.size.width, height: ratio * self.size.height)
        if UIScreen.main.responds(to: #selector(NSDecimalNumberBehaviors.scale)) {
            UIGraphicsBeginImageContextWithOptions(newSize, true, 0.0)
        } else {
            UIGraphicsBeginImageContext(newSize);
        }
        UIRectClip(clipRect)
        self.draw(in: clipRect)
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage!
    }
}
